### PR TITLE
fix clear history happening too often when setting is false

### PR DESCRIPTION
**Pull request by jharris4**
also clear history on disconnect from server when settings is false

also improve disconnect handling

### CODE REVIEW

File: src/shared/modules/dbMeta/dbMetaDuck.ts [hunk 1]
```diff
@@ -531,6 +534,7 @@ export const dbMetaEpic = (some$: any, store: any) =>
     .ofType(UPDATE_CONNECTION_STATE)
     .filter((s: any) => s.state === CONNECTED_STATE)
     .merge(some$.ofType(CONNECTION_SUCCESS))
+    .merge(some$.ofType(SWITCH_CONNECTION_SUCCESS))
```
Comment (oskarhane) [inline src/shared/modules/dbMeta/dbMetaDuck.ts:537]: Is this really needed?
I _think_ it's already triggered by the lines right above by the connection state going into `CONNECTED_STATE`.

File: src/shared/modules/dbMeta/dbMetaDuck.ts [hunk 2]
```diff
@@ -551,6 +555,7 @@ export const dbMetaEpic = (some$: any, store: any) =>
               .ofType(LOST_CONNECTION)
               .filter(connectionLossFilter)
               .merge(some$.ofType(DISCONNECTION_SUCCESS))
+              .merge(some$.ofType(SILENT_DISCONNECT))
```
Comment (oskarhane) [inline src/shared/modules/dbMeta/dbMetaDuck.ts:558]: Good catch, just make sure it works in desktop (which it should do).

File: src/shared/modules/dbMeta/dbMetaDuck.ts [hunk 3]
```diff
@@ -658,9 +663,15 @@ export const serverConfigEpic = (some$: any, store: any) =>
...
+            store.getState()
+          )
           store.dispatch(updateSettings(settings))
-          // settings must be updated in state for this check to work
-          if (!shouldRetainEditorHistory(store.getState())) {
+          const shouldRetainHistory = shouldRetainEditorHistory(
+            store.getState()
+          )
+          // if setting changed either prev/next was false so clear history
+          if (prevShouldRetainHistory !== shouldRetainHistory) {
```
Comment (oskarhane) [inline src/shared/modules/dbMeta/dbMetaDuck.ts:674]: It's not clear to me if we should clear the history here? 
Move it in / out of local storage, yes.
Remove it from runtime state, why?

File: src/shared/modules/dbMeta/dbMetaDuck.ts [hunk 4]
```diff
@@ -699,5 +710,14 @@ export const serverInfoEpic = (some$: any, store: any) =>
     })
     .mapTo({ type: 'NOOP' })
 
-export const clearMetaOnDisconnectEpic = (some$: any, _store: any) =>
-  some$.ofType(DISCONNECTION_SUCCESS).mapTo({ type: CLEAR })
+export const clearMetaOnDisconnectEpic = (some$: any, store: any) =>
+  some$
+    .ofType(DISCONNECTION_SUCCESS)
+    .merge(some$.ofType(SWITCH_CONNECTION))
+    .merge(some$.ofType(SILENT_DISCONNECT))
```
Comment (oskarhane) [inline src/shared/modules/dbMeta/dbMetaDuck.ts:717]: This already happens in https://github.com/neo4j/neo4j-browser/blob/f76ff9c731f62ecb606d715872d6aed96f273885/src/shared/modules/connections/connectionsDuck.ts#L581-L588 so maybe create a separate epic just to handle this case so we don't get any unexpected side effects.

**Review by oskarhane — COMMENTED** [review]
A bit too many changes for my taste: pulling in and listening on new action types in dbMetaEpic and listening on new events on the clearMetaOnDisconnect epic.
That's a bit risky in this situation imo.

I've commented a bit and real only real question is if we really should clear the history in other cases than when we're switching connection?

File: src/shared/modules/dbMeta/dbMetaDuck.ts (hunk 3 above)
Comment (jharris4) [reply src/shared/modules/dbMeta/dbMetaDuck.ts:674]: This is a fail safe for when the connection is switched to a different server without the connection events.

File: src/shared/modules/dbMeta/dbMetaDuck.ts (hunk 1 above)
Comment (jharris4) [reply src/shared/modules/dbMeta/dbMetaDuck.ts:537]: From our conversation earlier I thought they were separate & distinct events. I guess the best way to clear up that question is by performing the actions in desktop that trigger it.

**Review by OskarDamkjaer — APPROVED** [review]
Changes look good to me, thanks for changing to the .do operator and removing unused imports as per my comments.

**Review by oskarhane — APPROVED** [review]
LGTM

**Review by oskarhane — APPROVED** [review]
👯

**Pull request merged**